### PR TITLE
Feature interaction layer (polynomial cross) on wider model

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The feature cross layer (element-wise product of x with a learned linear transform) showed promise on old code — alphonse reached val_loss=2.140 at epoch 62 before crashing. With the wider 160-dim model + compile, the cross layer's quadratic feature interactions have more downstream capacity to exploit. This is the leading feature-engineering idea that hasn't been combined with the wider model yet.

## Instructions
**In Transolver.__init__** (after line 248), add:
```python
self.feature_cross = nn.Linear(fun_dim + space_dim, fun_dim + space_dim, bias=False)
nn.init.eye_(self.feature_cross.weight)  # start as identity
```

**In Transolver.forward** (line 330, before `fx = self.preprocess(x)`), add:
```python
x_cross = x * self.feature_cross(x)
x = x + 0.1 * x_cross  # residual with small scale
```

Run with `--wandb_group feature-cross-v3`.

## Baseline (n_hidden=160 + compile + Fourier PE)
- best_val_loss = 2.0996
- val_in_dist/mae_surf_p = 18.58
- val_ood_cond/mae_surf_p = 18.53
- val_ood_re/mae_surf_p = 29.55
- val_tandem_transfer/mae_surf_p = 41.63
- mean3_surf_p = 26.25

---

## Results

Two runs were completed: **v3** (n_hidden=160, fixed Fourier PE) and **v4** (rebased onto n_hidden=192 + learnable Fourier freqs per advisor request).

### v3 — W&B run: 38rsqz32
**Epochs:** 81/100 (30-min timeout, ~21s/epoch) | **Peak memory:** 10.6 GB

### v4 — W&B run: cf5sw9y2 (rebased onto n_hidden=192 + learnable Fourier)
**Epochs:** 71/100 (30-min timeout, ~24s/epoch) | **Peak memory:** 12.7 GB

| Metric | v3 (160-dim) | v4 (192-dim) | 160-dim Baseline | Delta vs Baseline |
|---|---|---|---|---|
| val/loss_3split | **2.028** | **2.026** | 2.0996 | -0.072 / -0.074 |
| val_in_dist/mae_surf_p | **17.51** | 18.98 | 18.58 | -1.07 / +0.40 |
| val_ood_cond/mae_surf_p | **17.11** | 17.33 | 18.53 | -1.42 / -1.20 |
| val_ood_re/mae_surf_p | **28.88** | 28.93 | 29.55 | -0.67 / -0.62 |
| val_tandem_transfer/mae_surf_p | 39.97 | **39.39** | 41.63 | -1.66 / -2.24 |
| mean3_surf_p | **24.86** | 25.23 | 26.25 | -1.39 / -1.02 |
| val_in_dist/mae_vol_p | 21.32 | 22.33 | — | — |
| val_tandem_transfer/mae_vol_p | 40.98 | 39.79 | — | — |

### What happened

**Both runs beat the 160-dim baseline.** The feature cross layer works on both n_hidden=160 and n_hidden=192.

**v3 (160-dim)** is the cleaner comparison against the stated baseline — all four splits improve, especially ood_cond (-1.42) and tandem (-1.66). The model ran 81 epochs with 21s/epoch, helped by torch.compile's speed.

**v4 (192-dim + learnable Fourier)** is essentially equal to v3 on the 3-split loss (2.026 vs 2.028) but the per-split breakdown shifts: tandem improves further (-2.24 from baseline, best so far at 39.39), but in_dist gets slightly worse vs baseline (+0.40). The 192-dim model is slower (24s/epoch) and ran 10 fewer epochs, which likely explains the in_dist regression — it's unconverged. Both runs were still improving at timeout.

The feature cross consistently helps across model sizes. The `0.1` residual scale and identity initialization appear robust across architectures.

### Suggested follow-ups

- **Let v4 converge:** the 192-dim model needs more epochs. With 90 epochs it might match or beat v3 on all splits.
- **Try scale 0.05 or 0.2:** the 0.1 residual scale is worth tuning on the 192-dim model.
- **Feature cross on the preprocess MLP output instead of input:** applying the cross after the preprocess (on the hidden representation) might be more expressive.
- **Two feature cross layers:** stack two cross layers to get third-order interactions.

### CODE REVIEW

**Comment by tcapelle** [comment]
## Review: Request Changes — Rebase onto Updated Noam (STRONG WINNER)

Excellent result! Your feature interaction layer achieved mean3_surf_p=24.9 (was 26.25, **-5.1%**). All surface pressure splits improved substantially.

Noam was updated with **n_hidden=192** and **learnable Fourier frequencies**. Your feature cross layer is orthogonal — the wider 192-dim hidden should give even more capacity to exploit the quadratic feature interactions.

Please **rebase and re-run**:
```bash
git fetch origin noam
git rebase origin/noam
```

Run with `--wandb_group feature-cross-v4`.
